### ISSUE TITLE
Scan process superslow, upredictable, no indicator, cannot close app

### ISSUE THREAD

**Issue by lighthunt** [whyboris/Video-Hub-App#512]
Moving from #296, since I think it's a current issues at least on my system.
Scan process is very slow in latest version (0ec441e) compared to v2.2.3 even with a single folder and relatively not many video files.
Some comments on your comments ;-) :

> 
> 
> Thanks for bug reports 🤝
> Some comments in random order:
> 
>     * `.jpg` file names are based off video file bitstreams (a bit from the front, middle, and back), so moving the video file doesn't change the "hash" - keeping the image file name the same 😁
> 
and that's how also search potential duplicates are identified easy by this hash, understand after some hints found when doing translation today, thanks ;-)

>     * (re)scanning multiple folders _should_ work without error - the scanning is a parallel process, and each file is added to a single queue which extracts metadata; all those files are added to a (different) singe queue which does the screenshot extraction
> 
>     * screenshot extraction (top right indicator) should persist until the thumbnail queue is done (so when you rescan and no video files appear, it should either not appear at all or go away swiftly).
> 
it's usually not there, or just blinks and disappears, but it still appears to be scanning

>     * initial scan for files should have a green indicator on top-left (next to number of files in the gallery) and should go away as soon as we're done collecting a list of all the files in all the directories the app is (re)scanning
> 
haven't noticed, but will observe next time
>     * I think I don't have a metadata extraction indicator, but as soon as one file has been added, it's now in the "thumbnail extraction" stage which should show the top-right indicator (the donut progress bar)
> 
> 
> After merging the WIP `3.0.0` branch into `main` I've not tested the app too much on the create hub/add folder/rescan/switch hub/rescan/close (etc) interactions. I'll likely discover those bugs when I try to play with the app more, but I _definitely_ appreciate reports like this. Please feel free to open a new issue for better visibility if you'd like.
> 
> As of now, I expect the app to not have bugs in a standard way of use: open, create new hub, add some more folders, exit. Re-open, have no bugs 😅 🤞 I've tested it with network drive, external HD over USB, and local HD.
> 
> I have not tested closing the app mid-scan, but I don't expect the app to break catastrophically 🤔 -- it should allow re-opening of the file and a re-scanning without problems.

The app didn't want to close, and when it did, the previously mentioned exception appeared:

> A JavaScript error occured in the main process
> Uncaught Exception:
> Errror: ECANCELLED: operation cancelled, scandir ....

After that, I didn't see any update in the vha2 file and I believe once the vha2 file disappeared (although this I can't reproduce anymore).

**Comment by whyboris** [comment]
Thank you for the bug report (and for opening a new issue for it)!

I'll need to test the current version more (on Windows and Mac and maybe my wife's computer too) to see if any weird bugs show up. As often is the case, _Works on my machine_ ™️ :trollface: 

I've yet to test the extraction speed against the older version. It seemed fast-enough on my PC, but my PC is souped up with an SSD (solid state drive), and I always extract stuff from other hard drives 🤔 

I did notice that the CPU goes sky-high while generating screenshots (but that's expected). I also think my video card was helping out somehow, though unsure 🤔 

For easy reference, if you don't mind, could you share your OS, RAM, and where your videos are / where VHA folder is (if it's same / different hard drives, network, SSD, or anything else you think is relevant)?

**Comment by lighthunt** [comment]
tested 2 laptops
old hp zbook gen 2 laptop: win7, ram 32GB, vha on local HDD (diskcryptor), video files on USB drives (veracrypt)
dell xps13, win10, 16GB, vha on local SSD, videos on local SSD or same USB drives

forgot to mention, when I tried to "reuse" filmstrips and thumbnails from v2 and copied them to a new v3 with (with that folder added), it didn't help (tested on zbook).
(disk encryption shouldn't matter)

**Comment by whyboris** [comment]
32GB 🤤 NICE!
Thank you for testing on multiple devices. My immediate thought is something about USB - parallel access of files maybe?

Are you using an external hard drive connected through USB, or a USB stick with some videos? (I want the app to work beautifully on both, but perhaps this is a clue about something. I tested with an external hard drive through USB and it worked well for me) 🤷 

Before `3.0.0` VHA did the scanning of files (file list), _then_ did meta data extraction, _then_ did screenshot extraction. Each in sequence, never concurrently.

`3.0.0` does the 3 steps in parallel. Before the initial scan is finished (now using `chokidar` library), it adds files for meta data extraction to a queue (using `async` library) which instantly tries to extract metadata one file at a time.

When adding the "delete old screenshots" functionality, everything ground to a halt, until I realized that the scanning was interfering with the deleting -- I promptly fixed it.

I don't see why generating a lists of files would conflict with metadata extraction, but perhaps on a (thumb-drive) USB parallel file access is particularly bad?

All ideas welcome 👂

**Comment by whyboris** [comment]
In a separate issue -- I've not yet spent time debugging what happens when you close the app mid-way through a scan. As far as I can guess about my code, it _should_ handle it well-enough, but I'll be empirically testing it later.

With the current setup, the app shouldn't freeze any more -- the "front end" of it should be responsive and fine -- it just listens for incoming new video data one by one (throttling adding it to the display after the first 20 videos/elements arrive) 🤔

**Comment by lighthunt** [comment]
Thanks for the note about parallelization. These are usually hardest types of issues :( 
(I'd guess parallel scanning of more "roots" makes sense, but the rest, not sure it's worth the trouble, but since already implemented let's see if it can be optimized/fixed). 

Past scenario details: I had 4 USB drives (500MB - 3GB) connected to zbook, 2 of them via passive USB hub, each having one root folder, so there could be some throttling and I cannot rule out they are very stable, they may sometimes dismount by veracrypt (but hey, this needs to be robust, so they are ok for testing)

My latest test to rule out HW issues and simplify my XPS (all on a single ssd, reliable).
1 root folder only, set for not watching, I clicked to manually rescan, rescan button animates for few seconds, the laptop fan spins (clearly scanning) "loading ... % " in window header apears, then disappears. (it sometimes cycles .. appears/disapears ... ).
From here (when either showing the loading or not), when I close the app (window close), I got the uncaught exception mentioned earlier.

![2020-08-21 12_40_15-](https://user-images.githubusercontent.com/14195401/90856149-847e3d80-e3ab-11ea-8f00-15109d2a4cd7.png)

Sorry I didn't measure the time, since I don't know whet if finish, but having the subjective feeling overall scanning is way slower than in v2 since I didn't have patience to wait ;-)

Didn't try usb thumbs, but I don't think it's relevant at this point.

If you have any more specific question I will try to reproduce should you need, good luck hunting.

**Comment by cal2195** [comment]
Sounds like it's related to #526

**Comment by cal2195** [comment]
In particular, that last error @lighthunt mentioned sounds like a symptom of my same issue! 👍

**Comment by whyboris** [comment]
My PR #541 may fix this, as I implemented the sage advice from @cal2195 in #526  🙇

**Comment by whyboris** [comment]
Hey @lighthunt -- if you have some time, could you try the current `main` branch again with your videos and the same settings that caused problems before and comment?

I strongly suspect at least some of the issues have been resolved with the merge of #526 🤞 

If you think all the problems are solved, feel free to close the issue 😁 or share what still is not 💯 🤝

**Comment by lighthunt** [comment]
@whyboris great!!!

Sorry I was not around lately, but I will definitely check this one during this week and let you know.

**Comment by lighthunt** [comment]
@whyboris although, I have run into some weird issues (I will report separately), the scan now became usable again and I think I can start using v3 again and hopefully test it a bit more ;-). Thanks! closing.
